IoT is internet of thing

With the advent of IoT, the attack surfaces on any given target have increased exponentially.


In this project we will build a machine learning model to classify netwrok IoT devices by type.

A dataset has 2 types which are iot_train.csv and iot_test.csv

In [1]:
# import pandas and os and read in the training and testing data
import pandas as pd
import os
training_data = pd.read_csv("data/iot_devices_train.csv")
testing_data = pd.read_csv("data/iot_devices_test.csv")

In [2]:
training_data

ack  ack_A  ack_B  bytes  bytes_A  bytes_A_B_ratio  bytes_B  ds_field_A  \
0      9      5      5   1213      743         0.713924      668           0   
1      9      5      5   1213      743         1.806874      668           0   
2      9      5      5   1213      743         0.103124      668           0   
3      9      5      5   1213      743         1.806874      668           0   
4      9      5      5   1213      743         1.806874      668           0   
..   ...    ...    ...    ...      ...              ...      ...         ...   
995    0      0      0    240        0         0.000000      240           0   
996    0      0      0    240        0         0.000000      240           0   
997    0      0      0    240        0         0.000000      240           0   
998    0      0      0    240        0         0.000000      240           0   
999    0      0      0    240        0         0.000000      240           0   

     ds_field_B  duration  ...  suffix_is_co.il  suffix_is_com  \
0             0    1.5756  ...                0              0   
1             0    0.6890  ...                0              0   
2             0    0.9852  ...                0              0   
3             0    1.5756  ...                0              0   
4             0    1.5756  ...                0              0   
..          ...       ...  ...              ...            ...   
995          64    1.4983  ...                0              0   
996          64    1.4979  ...                0              0   
997          64    1.4950  ...                0              0   
998          64    1.4993  ...                0              0   
999          64    1.4951  ...                0              0   

     suffix_is_com.sg  suffix_is_else  suffix_is_empty_char_value  \
0                   0               1                           0   
1                   0               1                           0   
2                   0               1                           0   
3                   0               1                           0   
4                   0               1                           0   
..                ...             ...                         ...   
995                 0               1                           0   
996                 0               1                           0   
997                 0               1                           0   
998                 0               1                           0   
999                 0               1                           0   

     suffix_is_googleapis.com  suffix_is_net  suffix_is_org  \
0                           0              0              0   
1                           0              0              0   
2                           0              0              0   
3                           0              0              0   
4                           0              0              0   
..                        ...            ...            ...   
995                         0              0              0   
996                         0              0              0   
997                         0              0              0   
998                         0              0              0   
999                         0              0              0   

     suffix_is_unresolved  device_category  
0                       0  security_camera  
1                       0  security_camera  
2                       0  security_camera  
3                       0  security_camera  
4                       0  security_camera  
..                    ...              ...  
995                     0           socket  
996                     0           socket  
997                     0           socket  
998                     0           socket  
999                     0           socket  

[1000 rows x 298 columns]

In [3]:
print(testing_data)

     ack  ack_A  ack_B   bytes  bytes_A  bytes_A_B_ratio  bytes_B  ds_field_A  \
0     38     20     18   14698     7814         1.135096     6884           0   
1     38     20     18   14698     7814         1.135096     6884           0   
2     38     20     18   14762     7814         1.124640     6948           0   
3     38     20     18   14762     7814         1.124640     6948           0   
4     41     20     21   14862     7994         1.163949     6868           0   
..   ...    ...    ...     ...      ...              ...      ...         ...   
895  264    116    148  212053   202036        20.169310    10017           0   
896   24     13     11    7749     5364         2.249056     2385           0   
897   20      9     11    7091     5336         3.040454     1755           0   
898   30     14     16    7882     5789         2.765885     2093           0   
899  294    147    147  209972   197919        16.420724    12053           0   

     ds_field_B  duration  

That means the data contains 298 features

In [4]:
# We will create a training and testing dataset where the target is the device category:
x_train,y_train = (
    training_data.loc[:, training_data.columns != "device_category"].values,
    training_data["device_category"],
)

x_test,y_test = (
    testing_data.loc[:, testing_data.columns != "device_category"].values,
    testing_data["device_category"],
)

The device categories are security camera, TV, smoke detector, thermostat, water sensor, watch, baby monitor, motion sensor, lights and socket

In [5]:
# We need to encode the class categories into numerical form:
from sklearn import preprocessing
le= preprocessing.LabelEncoder()
le.fit(training_data["device_category"].unique())
y_train_encoded = le.transform(y_train)
y_test_encoded = le.transform(y_test)

In [6]:
# We will instantiate XGBoost classifier
from xgboost import XGBClassifier
model = XGBClassifier()

ModuleNotFoundError: No module named 'xgboost'

In [ ]:
# We need to train and test the xgboost classifier
model.fit(x_train, y_train_encoded)
model.score(x_test, y_test_encoded)

0.6622222222222223

An important motivation for this project is that we cant reply on the IP address as an identifier of the device, since this value can be spoofed. Consequently we would like to analyze the traffic's high-level data, that is the metadata and traffic statistics, rather than content to determine wheterh the device belongs to the network. 

1.   Step 1:We begin by reading in the training and testing dataset on to featurize these and perform a quick data exploration step by observing the classification labels
2.   Step 2: We feed these into our classifier.

1.   Step 3: we convert these categorical labels into numerical ones to be used to train our machine learning classifier.

1.   Step 4: We having featurized the data

1.   Step 5: We instantiate train and test an xgboost classifier, obtaining a score of 0.66 on the testing set.


The baseline of randomly guessing between 10 would yeild an accuracy of 0.1. The XGBoost classifier trained here attains an accuracy of 0.66 suggesting that it is indeed a promising approach to classify IoT devices successfully based on high-level traffic data.



